Add a document describing options for managing storage

berrange opened the pull request:
This document describes the current approach of using PVCs or network block
devices for persistent storage. It then illustrates how this concept can be
extended to enable the notion of ephemeral storage to be supported. Finally
it describes at a very high level, how the PVCs can be populated using a number
of techniques.

stu-gott commented on docs/storage-config.md:
```diff
@@ -0,0 +1,204 @@
+with one or more LUNs. The POD would run a self-contained service that just
+pulls down, one or more, named images from a container repository, extracting the
+disk image files (raw / qcow2 / whatever) from each image. The extracted images
+that then exported by the iSCSI server. The ISCSI LUNs can be directl referenced
```
typo: "directly"

fabiand commented on docs/storage-config.md:
```diff
@@ -0,0 +1,204 @@
+provide ephemeral storage that is discarded when the virtual machine is deleted
+from the node.
+
+The host needs to have somewhere to storage the local overlays / files. This is
```
s/storage/store/

fabiand commented on docs/storage-config.md:
```diff
@@ -0,0 +1,204 @@
+this purpose. If the PVC backing store is only accessible for a single host,
+then when performing migration, QEMU block storage migration feature will need
+to be enabled. If the PVC references a networked filesystem that can be mounted
+on multiple hosts, then VMs can migrate with no need to copy emphemeral disks.
```
Wouldn't we need to copy data, if we used a network storage with a local caching copy mode?

fabiand replied:
… or maybe not if the volume where the cache is kept is shared between the hosts?

berrange replied:
Whether the network storage driver has local caching or not is tangential to this choice. It can potentially improve the performance of any of these options.

mykaul commented on docs/storage-config.md:
```diff
@@ -0,0 +1,204 @@
+
+#### Direct storage reference
+
+The storage server can be referenced directly, for example, as an iSCSI server
```
Is the connection from qemu or the host? (I assume from qemu, which is a bit different feature-wise and management-wise from the host)

berrange replied:
This is from QEMU, host is not involved - the latter would be block device based as described in previous paragraphs

fabiand commented on docs/storage-config.md:
```diff
@@ -0,0 +1,204 @@
+would need to be used to tag them for different usage types. This might be useful
+if a cluster needs to provide ephemeral storage with differing performance grades
+or costs (eg one volume provides cheap & slow rotating rust, while another volume
+provides expensive & fast sold state).  The VM spec disk configuration would be
```
s/sold/solid/

fabiand replied:
s/  / /

fabiand commented on docs/storage-config.md:
```diff
@@ -0,0 +1,204 @@
+if a cluster needs to provide ephemeral storage with differing performance grades
+or costs (eg one volume provides cheap & slow rotating rust, while another volume
+provides expensive & fast sold state).  The VM spec disk configuration would be
+able to optionally include a reference to the usage type it desires, to allow
```
Reminds me of storage classes https://kubernetes.io/docs/concepts/storage/persistent-volumes/#class

mykaul commented on docs/storage-config.md:
```diff
@@ -0,0 +1,204 @@
+
+This would also work for RBD, NBD, GlusterFS, and HTTP(s) protocols.
+
+#### Indirect storage reference
```
Any performance (latency?) implications here?

berrange replied:
This is no different to the previous option - both end up using the in-QEMU storage driver.

fabiand commented on docs/storage-config.md:
```diff
@@ -0,0 +1,204 @@
+based. Using QEMU's direct network backend drivers, it is possible to directly
+access the storage for any PV that is backend by a network block device (in
+particular the iSCSI and RBD k8s backends).
+
```
I'd enclose it in preformatted text thingy studd, aka ```yaml…

fabiand commented on docs/storage-config.md:
```diff
@@ -0,0 +1,204 @@
+
+Thus the libvirtd container may contain
+
+  annotations:
```
I'd also add the yaml format thingy … ```yaml

fabiand commented on docs/storage-config.md:
```diff
@@ -0,0 +1,204 @@
+
+When configuring the VM spec disk
+
+    disks:
```
same here

stu-gott commented on docs/storage-config.md:
```diff
@@ -0,0 +1,204 @@
+
+### HTTP upload
+
+A "HTTP upload" POD could be created that references the PVC that needs to
```
Presumably we would only want this pod to run temporarily--long enough for the user to upload their intended image, then it would exit. It might be worth mentioning that.

fabiand commented on docs/storage-config.md:
```diff
@@ -0,0 +1,204 @@
+whose backing store refers to the master PVC.
+
+
+## Populating PVCs
```
I'd move this section into a separate proposal, to keep this one focused on the copy modes.

berrange replied:
I want to keep it here actually, because it is relevant to an enhancement i'm adding to the disk source options

berrange replied:
This is relevant to a further update I just made to the doc, which lets you refence a container spec  as the disk source: to auto-create PVCs at VM startup time.

mykaul commented on docs/storage-config.md:
```diff
@@ -0,0 +1,204 @@
+There is no general requirement that any disk source be used exclusively by a
+single virtual machine. There are three permissible access modes
+
+* Write-exclusive: a single virtual machine is writing to the storage. No
```
I wonder if we are missing a mode of single writer multiple readers?

berrange replied:
In terms of data consistency requirements this is identical to the multiple-writers case, because the readers have to cope with data changing behind their back.

fabiand commented:
/cc @michalskrivanek @aglitke

berrange commented:
I pushed an updated version of the doc which addresses the comments, but also illustrate how we could define a general purpose VM spec disk source which lets virt-handler run an container to auto-create the real disk source. This should better address the scenario David was trying to achieve by integrating docker container support into virt-handler, but without virt-handler needing to know about docker support directly.

berrange commented:
This also  adds an example describing how to address the clear containers use case, by having a pod that can auto-create qcow2 disk images populated from an existing docker image.

fabiand commented on docs/storage-config.md:
```diff
@@ -0,0 +1,267 @@
+
+Rather than having the storage volume pre-created upfront, it may be useful to
+have a way to dynamically create storage backends. To achieve this, it is
+possible to define a storage source which defines a container. The virt-handler
```
What you are describing here is probably covered by dynamic provisioning in Kubernetes, see https://kubernetes.io/docs/concepts/storage/persistent-volumes/#dynamic - or at least it looks close enough on a first glance.

berrange replied:
The key difference with my suggestion is that this is entirely under the control of the tenant user. They don't have to rely on the cloud admin to have configured anything, whether in virt-handler, or k8s persistent volume / storage class setup. It is also flexible, as the user doesn't even need to create PVCs - they can just return an ISCSI server address directly if desired.

aglitke replied:
I think this is a really neat idea, but I think we should strive to integrate as tightly with k8s as possible.  Doing storage "out of band" as you suggest leaves kubevirt vulnerable to future breakage and diverges the project from mainline storage feature development.  Kubernetes has a flexvol extension interface which allows you to write your own storage providers and provisioners.

berrange replied:
Flexvol doesn't really fit in with way we need to consume storage. With a flex vol backed PVC, you just end up with a mount point inside the POD, which we can't do anything with. We need to get the address of the underlying network storage (iscsi/rbd/nbd/etc), without it ever being mounted anywhere in the host.

berrange replied:
Also as mentioned above, IIUC, PVs & storage classes are something that has to be managed by the cloud administrator. What we're describing here is a mechanism for tenant users to dynamically expose storage for VMs, using PVCs concepts they already have access to. Basically its a short cut to avoid the tenant needing to manually create a PVC before hand

aglitke replied:
FYI, block storage support is being actively worked on in the k8s storage community.  Until this lands we will certainly have issues consuming the native API and might need a workaround. 
 See https://docs.google.com/document/d/1XeNFxc89C54psYqz4RErk1xcso0wMrKtSfWqI6POvaU

Also, I consider it a benefit to use the cloud provided storage resources rather than BYO.  In this way, the storage available to all applications (kubevirt or otherwise) is managed the same way and from the same place.  IIUC, there should be no need to pre-create PVs as long as dynamic provisioning is configured for the storage class being requested by the PVC.

berrange replied:
This is really not duplicating storage class concept, rather it is complementary to it. The k8s dynamic provisioning doesn't do anything more than creating the volume, it does nothing for actually getting interesting content into that volume. The concept proposed here could actually use the k8s StorageClass facilty as part of its impl to create the volume that it then populates with data from a qcow2 file, or some other arbitrary source. This also allows for things that cannot be done with the k8s storage concepts - currently those are entirely under the control of the cloud admin. This gives the tenant user a mechanism to have direct control, if what the cloud admin provider is insufficient for their needs. 

Directly comparing with containers is also not entirely useful as there are reasons tenants can't directly create their own storage. Allowing the tenant user to create arbitrary storage that is then mounted on the host is a serious security risk - the tenant could create a malicious filesystem that crashes the host kernel. This is safe to do for VMs though, because the thing consuming the content is the guest OS, the host is never involved.

rmohr replied:
A user can always ask for disk space via a PVC, upload there whatever he likes and serve a disk to the VM with it's own Pod in many ways. Why do we need extra specifications for such a `free-form` thing? Further there is already a clear guide on how to implement your own dynamic provisioner based on the `StorageClass`, which works completely different than what you are proposing here (https://github.com/kubernetes/community/blob/master/contributors/design-proposals/volume-provisioning.md). I can definitely see us implementing one or more of such provisioners for some advanced features.

berrange replied:
Only using PVCs directly is unnecessarily restrictive on the user. The user could have another VM already running that is using a PVC, and then exporting storage via some other technology - for example they might export an NFS share from a VM, backed by a PVC and wish to configure other VMs to use files in the NFS volumes as virtual disks. Or they may have something that is dynamically creating NBD volumes containing the data for a CDROM image to act as backing storage for cloud-init. The use of the StorageClass is certainly possible, but it is unnecessarily restrictive because that's controlled by the cloud admin, not the tenant user. This gives the tenant user more flexibility in how they deal with storage for the VM, so they're not entirely reliant on the cloud admin for everything.

fabiand commented on docs/storage-config.md:
```diff
@@ -0,0 +1,267 @@
+allowing existing docker images to be directly consumed as if they were disk
+images. Thus tenant users would never need to deal with qcow2 images directly,
+they can use normal docker tools for building images & let kubevirt boot them
+as-is.
```
This is tricky in the detail. I played a little bit with it in https://github.com/fabiand/vocker/

berrange replied:
IIUC, your approach there is different to what I describe. Your repo appears to taking the orginal Dockerfile and then directly building a disk image from it. What i'm describing here runs later - it just pulls down the layers from the docker registry, and puts the content into a qcow2 image. This works pretty well - we've implemented this previously for libvirt http://libvirt.org/git/?p=libvirt-sandbox.git;a=blob;f=libvirt-sandbox/image/sources/docker.py;hb=HEAD

fabiand replied:
Yes - That's true.

But there was - let me find it - a branch where i transformed docker images into qcows (not layer wise) and then booted them.

davidvossel replied:
This adapter concept is a little convoluted to me. Approaching it from the user perspective will help give me clarity.

A user has a vm image, they want that image in the docker registery in a way that it can be consumed by this "container image to disk adapter". What are the steps the user has to go through to build the container image?

berrange replied:
The Dockerfile the tenant would use would just contain 

   ADD myfedora25.qcow2 /vmimage 

build this image & upload it with name 'myfedora25'

In the disk source, they would provide

  source:
    container:
      image: kubevirt-container-image-serve
      env:
         - name: KUBEVIRT_IMAGE_SOURCE
           value: myfedora25

Assuming that populated a PVC with the image, the REST endpoint would spit out

  source:
    pvc:
      claimname: some-pvc-name

davidvossel replied:
got it. we're mostly on the same page.

From what I gather, the only place we differ is in whether a HTTP endpoint coordinates provisioning the PVC using the container image, or if virt-controller coordinates that bit.

davidvossel commented on docs/storage-config.md:
```diff
@@ -0,0 +1,267 @@
+would create an instance of this container, which would be required to expose a
+HTTP service with a well-known URI eg https://$POD-IP/volume
+
+Making a PUT request to this URI would trigger creation of the volume, and a
```
I'm fairly certain we don't need an http endpoint for this, and virt-handler shouldn't be defining containers to do anything.

When a VM spec comes in, *virt-controller* defines the virt-launcher pod. As a reference, this virt-launcher pod is the one that would have the "angle" container in your original design.  If *virt-controller* see's these that dynamic storage needs to be provisioned for a VM, then *virt-controller* can coordinate whatever it means to provision that storage before the controller allows the VM to be launched.

When a VM is deleted, virt-controller will handle executing whatever jobs need to happen to cleanup dynamically created persistent storage associated with the VM.

When local ephemeral storage disks are assigned to a VM, it would be wise for us to utilize the virt-controller pod to provide those disks (by adding additional containers to the pod to serve disks). This means we get automated cleanup of the disks without having to track storage lifecycles on the local node.

berrange replied:
If not using a HTTP endpoint, what channel would you suggest for the container to return the desired disk source yaml ?

Having the virt-controller run the PODs for populating dynamic disk sources is certainly a viable option, if we're happy considering virt-controller to be a required part of the startup flow.  I was still thinking from the POV of eliminating virt-controller. Whether the disk create lives in the controller or handler though is a fairly minor distinction, as long as we think the overall concept of how the disk is expressed is sane.

aglitke replied:
If using PVCs with a flexvol driver that supports dynamic provisioning I think k8s would handle volume lifecycle itself.

berrange replied:
As above, the flexvol concept doesn't really work for what we need.

rmohr replied:
@aglitke while the flexvol driver might be too specific to reuse it for our purposes, I could see that we might have something similar in the future. It could be an extension point, to provide a disk in a generic way, with logic completely outside of KubeVirt, like the flexvol allows for pure k8s user. 

For some use-cases  the flexvol driver could even be the one used, to provide a disk. For instance, the logic in #325 could also be used, to serve a disk via a local iscsi connection, which was originally brought in by a custom implementation of a flexvol mount. I would not want to pursue something like this right now, though ...

davidvossel commented:
In general I agree with a lot of the concepts in this pull request. The idea of dynamically populating generic block storage with vm images from some registry is a good one.

aglitke commented on docs/storage-config.md:
```diff
@@ -0,0 +1,267 @@
+or costs (eg one volume provides cheap & slow rotating rust, while another volume
+provides expensive & fast solid state). The VM spec disk configuration would be
+able to optionally include a reference to the usage type it desires, to allow
+virt-handler to map the disk to the particular volume mount.
```
This seems like reinventing k8s storage classes to me.

aglitke commented on docs/storage-config.md:
```diff
@@ -0,0 +1,267 @@
+populate some PVCs with persistent data, or they could be used as part of the
+dynamic disk source described above, to create storage on-demand during VM
+startup.
+
```
I wonder if the below implementations could all be flexvol drivers in a common repository to enable code sharing.  Each of these is unique in how it presents a certain backend storage to the pod but they all would want to share a common framework for accessing/provisioning volumes on the supported storage backends.

rmohr replied:
@aglitke see also #325 for a possible implementation, which does not involve PVCs or anything like that, for making use of docker images.

fabiand commented:
On _Dynamic storage reference_. There could be a service, as described, which provides a REST endpoint to upload data - or, similar to described - other mechanism to generate content.

Regardless how the data is produced (upload, conversion, retrieval from another source [i.e. docker]), this service could
1. create a PVC
2. the cluster creates a PV matching the request
3. the service binds the PVC
4. the service populates the PVC with data
5. the service unbinds the PVC
6. the service returns the PVC data

Authentication might be an ssue here, as the service is obviously doing something on behalf of a user.

This would not leverage the existing dynamic provisioning features of Kubernetes, but would still provide the dynamic aspects described in the proposal.

To me this is independent of the suggested _copy modes_, which sounds like nice idea. To me there are questoins i.e. how it can be used.
I.e. if it's implicit (i.e. done by kubevirt) that a PV is requested, if a copy mode is used, to hold the overlay. Or if it's an explicit parameter to specify a PV or storage to be used for the overlay.

So all in all enough content for more that one proposal, and would be nice if we could continue with it :)

rmohr commented on docs/storage-config.md:
```diff
@@ -0,0 +1,267 @@
+Each of the volume mounts provided for emphemeral storage would need to have a
+libvirt storage pool associated with them. To create the ephemeral disks, the
+virt-handler would thus use the storage pool APIs in libvirtd to create disks
+whose backing store refers to the master PVC.
```
I guess, this storage pool has the logic to do the cleanup of the temporary files?

berrange replied:
There's no notion of temporary files in libvirt APIs. It just provides a mechanism - its upto the app to decide when to clean things up, if it considered them temporary files

rmohr commented on docs/storage-config.md:
```diff
@@ -0,0 +1,267 @@
+use of this feature would require the cloud admin to manually create libvirtd
+POD specs for each compute node, instead of using a DeamonSet resource. It is
+anticipated that either DaemonSet would be enhanced eventually, or replaced by
+a more advanced resource that can dynamically assigns PVCs to each POD.
```
There is no special binding between a pod and a PVC. A user just creates and deletes PVCs when needed. virt-handler has to take the access modes into account, and once the PVC is not needed anymore, one deletes it and the dynamic provisioner will recycle the associated PV. There is not much more involved.

berrange replied:
The point being made here is that when you are having PODs created indirectly via a daemon set, and those PODs require storage volumes, there's no way to have a distinct storage volume provisioned or allocated for each POD. There might be cases where it is ok that every POD created gets the same shared storage volume, but equally there's cases where that's certainly not ok. This is a known limitation of the DaemonSet controller today for containers that is likely to be addressed in future in k8s.

rmohr commented on docs/storage-config.md:
```diff
@@ -0,0 +1,267 @@
+The 'usage-type' attribute refers to the 'kubevirt.io/sotrage.usage-type'
+annotations provided by the libvirtd POD. Either all libvirtd PODs would need
+to support the same annotations, or the scheduler will need to be able to only
+place VMs on the nodes where the libvirtd POD has the neccessary annotations.
```
Such annotations can be placed on the node directly, no need to place them on a specific pod. The scheduler can take them into account. VMs have a `nodeSelector` section, which is mapped to the `nodeSelector` section of a Pod. That is taken into account.

rmohr commented on docs/storage-config.md:
```diff
@@ -0,0 +1,267 @@
+
+### HTTP upload
+
+A "HTTP upload" POD could be created that references the PVC that needs to
```
PVCs are backed by PVs. If we don't use PVs, there is no advantage in having a PVC. Just "somehow" provide your image, and reference it with the VM equivalent of a Pod volume. You can use Pods and Services, like for any arbitrary stuff, to provide something like this via k8s.

berrange replied:
Err what's written here *is* making use of PVs. It is describin a way to get data into a PV associated with a PVCs

rmohr replied:
Good.

rmohr commented on docs/storage-config.md:
```diff
@@ -0,0 +1,267 @@
+
+### Container repository copy
+
+A "contanier repository copy" POD could be created that references the PVC that
```
Again, without PVs, it does not make much sense to use PVCs at all. They only exist to decouple the storage logic (complex, provider specific and needs a lot of privileges) from the need to express the requirement for a volume by the user.

berrange replied:
As above, what's describing is using PVs

rmohr commented on docs/storage-config.md:
```diff
@@ -0,0 +1,267 @@
+copy is complete the POD would exit
+
+
+### Container repository stream
```
That seems to be the one described in #325. It can be done completely local, managed by virt-handler. Since it came up in #325 this is special enough, when taking into account how docker images are distributed, that we  can handle that purely on the node. It is really more like the `image` field in a PodSpec, rather than a managed volume.

berrange replied:
The whole point of this design proposal is to avoid having to do special cases in this way, particularly since it is far from clear that #325 is a desirable approach in general.

rmohr replied:
And this concept does not fit into k8s. There are a few things you can do, when you want dynamic provisioning of PVCs, which fit much better into k8s. Moving this kind of generalized dynamic provision as our main entrypoint to virt-handler does not make sense in a k8s context.
The layer in virt-handler you introduce with your rest API does two things:
 * Things which virt-handler can do best directly  are moved away from virt-handler (e.g. copying an image from somewhere to the local machine), 
 * Things which are better done on the cluster side don't fit good into virt-handler flows, are now at least indirect the responsibility of virt-handler (e.g. triggering the run of a provision pod through the above rest service).

Like the kubelet, virt-handler is and should also in the future be designed, to either do host-related things directly ( everything host related: managing the VM object, mounting, copying, ... ), or wait until preconditions are not met (e.g. if a disk not provisioned, it should just try again later).

rmohr commented on docs/storage-config.md:
```diff
@@ -0,0 +1,267 @@
+Making a PUT request to this URI would trigger creation of the volume, and a
+a DELETE request would naturally delete it. The response to the PUT request
+would be a YAML document following the 'iscsi' or 'pvc' schema for the VM
+spec disk source. eg
```
I really don't see much value in this rest interface (per cluster/per node?). Either the user use PVCs and  the corresponding provisioners based on storage classes, or does her own thing, by just using Pods and Services, like any other custom k8s application. In the second case, it is completely out of the scope of KubeVirt, we just provide the gap, by allowing VMs to directly connect to storage (direct reference).

berrange replied:
The point is to tie the lifetime of the storage object, to the lifetime of the VM, and todo that you want it controlled by either the virt-handler or virt-controller during the VM startup procedure. This allows the storage to be created based on metadata associated with the VM that is being started. For example, with cloud-init, one way of providing data to it is via a CDROM device. The image for that CDROM needs to be generated & populated with data associated with the VM and should only live for as long as the VM is alive. The REST interface is just a mechanism to be used by either virt-controller or virt-handler to do the integration. In particular they don't know what type of storage will be setup - whether its a iSCSI vol, an NBD vol, a PVC (which must then be further resolved). The response to the REST API call provides the DiskSource document that describes the vol just created, to be used for the rest of the VM create process.

rmohr replied:
Definitely not virt-handler. virt-controller could help. There are differnet ways, but none of them require such a REST api on the virt-handler side, which like stated further down in another comment, moves responsibilities to the node, which are not node-related.

So there are different ways, how such flows can be enabled in a more k8s like approach:

1) Dynamic provisioners (with storage class)
2) full reedom, by doing your completely own storage with your own services and pods
3) Integrating e.g. a provision container into the VM spec, which will force virt-controller to create and provision the appropriate PVCs before the VM is even scheduled. Also note that a potential required delete of such a storage, after a VM stopped, would be done by a controller, not virt-handler.

Possible example:

```yaml
apiVersion: kubevirt.io/v1alpha1
kind: VM
metadata:
  name: testvm
spec:
  domain:
    ...
  storage:
    recycleStrategy: deleteOnVMStop
    template:
      ... my provisioner pod template ...
```
However, there might be reason, why we would still want to completely keep the storage lifetime and the VM spec separated. In this case the next option would also be interesting.

4) Create something like a Disk TPR, which contains a provisioner pod template:

```yaml
apiVersion: kubevirt.io/v1alpha1
kind: ProvisionedDisk
metadata:
  name: testvm
spec:
  pvc:
    template:
    ... my pvc template ...
  provisioner:
    template:
      ... my provisioner pod template ...
status:
  phase: Bound
```
Bound in this case could mean for virt-handler, PVC is ready to go and provisioned, so resolve it and use it.


However, the main point is, that whenever we want to give the user such tools, we have to think about letting controllers do the cluster wide stuff and let the controller object status represent, when the resource is ready. virt-handler can then just consume it. Other things can be solved on the node. They should be done by virt-handler directly. A special rest api which virt-handler would query, does absolutely not make sense in any way. 

Something which in my eyes is a very good decision tree, on how to integrate with k8s is: First think about what the user (operator) want's to solve and what the steps are, then think about the scope (cluster vs node). Then, if it is a cluster scope, think about a controller which can do the steps for the operator.

Finally, there might still be the need for management applications, which want to use kubevirt, to provide something like an  upload rest-api to the user. All they would have to do, is to write their upload application, deploy it to k8s, configure a controller like outlined in (3) and (4) with the right pod template (to copy uploaded images to the right spot in a PVC, or transform the PVC into the image), an they are done.

berrange commented on docs/storage-config.md:
````diff
@@ -0,0 +1,267 @@
+    lun: 2
+```
+
+## Access modes
````
Access mode is relevant whether we use PVCs/PVs, or directly reference storage via ISCSI / NBD / etc.

rmohr replied:
Sure, but afaik the qemu process can't enforce an access mode on the storage. The storage has to provide that feature, right?

berrange replied:
QEMU can enforce access mode on storage that is file based, but not currently on network based storage. It would be technically possible to enforce it on iSCSI based storage using SCSI reservations, but that's not implemented right now.  Libvirt also has ability to perform locking on disks volumes, and has a number of options in this area. Again currently, it doesn't do it for network based storage, but it is technically possible to extend it. So in general access mode concept makes sense.
